File: src/java/Decipher.java
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.RandomAccessFile;
import java.io.IOException;
import java.util.*;
import java.lang.*;

public class Decipher {
	public static HashMap<String, Integer> dictionary;
	public static HashMap<String, Vector> patterns;
	public static String globalCipher;
	public static int wordFreqCount;
	public static Vector<Mapping> possibleKeys;
	public static Object[] wordsByLength;
	
	public static void main(String[] args) throws IOException {
		/**
		 *  Reading Source Files including your data structures
		 */
		
		/*******************************************************
		 You should do some pre-processing on given text sources and store them in a file so as to not do pre-processing each time.
		 In this part you should read the files and fill the required data structures, like arrays, and assign values to some variables that you will
		 use for finding the plaintext.			
		********************************************************/
		dictionary = new HashMap<String, Integer>();
		patterns = new HashMap<String, Vector>();
		possibleKeys = new Vector<Mapping>();
		wordsByLength = new Object[19];
		/**
		 *  18 is the max word length in sources 1-3
		 */
		for(int i = 0; i <= 18; i++)
			wordsByLength[i] = new HashMap<String, Integer>();
		
		wordFreqCount = 0;
		
		/**
		 * Dictionary pre-processing
		 */
		if(!loadDictionary("dictionary.txt")) {
			parseFile("source1.txt");
			parseFile("source2.txt");
			parseFile("source3.txt");
			
			outputDictionary();
			
			loadDictionary("dictionary.txt");
		}
		
		// TODO: Add source pre-processing for storing plain-text sources to validate deciphered messages
		
		/**
		 *  Reading Input File, Deciphering and Writing to outputFile	
		 */
		List argz = Arrays.asList(args);
		Iterator it = argz.iterator();
		String inputFileName = it.hasNext() ? (String)it.next() : "";
		Long timeLimit = it.hasNext() ? Long.parseLong((String)it.next()) : 6000;
		File inputFile = new File(inputFileName);
		File outputFile = new File("Output.txt");
		String cipherText;
		RandomAccessFile inputFileRAF = new RandomAccessFile(inputFile,"r");
		BufferedWriter outputWriter = new BufferedWriter(new FileWriter(outputFile));
		
		/**
		 * Parse cipherText from file 1 line at a time	
		 */
		while ((cipherText = inputFileRAF.readLine()) != null) {
			/**
			 * Reading the cipherText from file
			 */
			
			cipherText = cipherText.toUpperCase();
			//Replace anything that is not A to Z or space with a space character
			cipherText = cipherText.replaceAll("[^A-Z\\s]", " ");
			//Replace all instances of more than one consecutive whitespace character with a single space character
			cipherText = cipherText.replaceAll("\\s+", " ");
			//Replace last whitespace with nothing
			cipherText = cipherText.replaceAll("$ ", "");
			
			String plainText = "";
			
			// Begin timer for deciphering text
			long startTime = System.currentTimeMillis(); // Start time
			long current = System.currentTimeMillis();
			
			do{
				plainText = beginDecipher(cipherText);
				current = System.currentTimeMillis();
				if(!plainText.equals(""))
					break;
			}while(current - startTime < timeLimit);
			
			
			//
			/************************************************************
			
			plainText = FindPlaintext(cipherText);
			
			In FindPlaintext function, you should somehow loop and check if the time that you spent on this cipherText
			is greater than a time limit or not. This means you should come up with a time limit for each deciphering operation. 
			You can decide it by testing on some cases, like for n = 50 TimeLimit = 3 min; n = 100 TimeLimit = 2 min, so on... 
			Your FindPlaintext function should be like this: 
			
			String FindPlainText(String cipherText){
				do{
					FindPlaintextOperation();
					long current = System.currentTimeMillis();
				}while(current - startTime < timeLimit)
				
			}
			
			*******************************************************************/
		
			long endTime = System.currentTimeMillis(); // Finish time
			long totalTime = endTime - startTime; // TotalTime calculation for finding a plaintext
			
			/* Printing output - Output Format: "Plaintext,Time(milli sec.)\n"
			   Each line in outputFile will consist of one <plaintext,time> pair
			   If no plaintext has been found, then replace plaintext with "?" mark. */			
			
			//if(solutionsOpt == 1)
			//	printAllPossibleMatches(cipherText);
			
			if(plainText.equals("")){	// if no plainText has been found for this cipherText
				
				outputWriter.write("?"+","+totalTime+"\n");
				System.out.println("?"+","+totalTime+"\n"); // To see that your code doing some progress,
															// please print out the pairs on the screen.
			}
			else{	// if you have a plainText for this cipherText
					
				outputWriter.write(plainText+","+totalTime+"\n");
				System.out.println(plainText+","+totalTime+"\n"); // To see that your code doing some progress,
																  // please print out the pairs on the screen.
			}			

		}

		inputFileRAF.close();
		outputWriter.close();
		System.out.println("I'm done...");
	}
	
	/**
	 * printAllPossibleMatches()
	 * @param String cipher
	 * @return void
	 * @function Prints current possible key matches for given cipher 
	 */
	public static void printAllPossibleMatches(String cipher){
		Iterator it = possibleKeys.iterator();
		System.out.println("Possible matches:\n");
		
		while(it.hasNext()) {
			String plainText = decrypt((Mapping)it.next(), cipher);
			System.out.println(plainText);
		}
	}
	
	/**
	 * beginDecipher()
	 * @param String cipher
	 * @return String plainText
	 * @function Wrapper function for Decipher method to initialize deciphering process
	 */
	public static String beginDecipher(String cipher) {
		globalCipher = cipher;
		String plainText = "";
		
		// Split ciphertext into string array
		//String[] cipherWords = cipher.split(" ");
		String[] sortedCipherWords = cipher.split(" ");
		// Sort by length
		Arrays.sort(sortedCipherWords, new StringLengthComparator());
		Stack<String> words = new Stack<String>();
		Mapping finalMap = new Mapping();
		
		for(int i = sortedCipherWords.length-1; i >= 0; i--)
			words.push(sortedCipherWords[i]);
		
		// TODO: words stack error-prone if empty - Option 1: Add conditional wrapper to encapsulate word pattern processing; Option 2: Fix decipher() logic to encapsulate word, pattern, candidate iteration 
		String word = words.pop();
		String pattern = pattern(word);
		//System.out.println(word + ": " + pattern);
		Vector candidates = (Vector) patterns.get(pattern);
		Iterator it = candidates.iterator();
		while(it.hasNext())
		{
			String s = (String)it.next();
			finalMap = new Mapping(word, s);
			
			finalMap = decipher(finalMap, words);
			if(validKey(finalMap))
				break;
		}
		
		if(!validKey(finalMap))
		{
			Iterator it1 = possibleKeys.iterator();
			while(it1.hasNext())
			{
				Mapping possibleMap = (Mapping)it1.next();
				if(evaluate(finalMap) < evaluate(possibleMap))
				{
					//System.out.println("Old Key:" + decrypt(finalMap, cipher));
					finalMap = possibleMap;
					//System.out.println("New Key:" + decrypt(possibleMap, cipher));
				}
			}
		}
		
		// Chomping for calculating accuracy
		plainText = decrypt(finalMap, cipher);
		/*StringBuilder s = new StringBuilder();
		List lst = Arrays.asList(plainText.split(" "));
		Iterator it2 = lst.iterator();
		while(it2.hasNext())
			s.append(it2.next());
		plainText = s.toString();
		*/
		return plainText;
	}
	
	/**
	 * decipher()
	 * @param Mapping currentMap
	 * @param Stack<String> words
	 * @return Mapping currentMap
	 * @function Recursively deciphers correct Mapping
	 */
	public static Mapping decipher(Mapping currentMap, Stack<String> words){
		if(words.empty())
			return currentMap;
		
		Mapping preservedMapping = new Mapping(currentMap);
		String word = words.pop();
		String pattern = pattern(word);
		//System.out.println(word + ": " + pattern);
		Vector candidates = (Vector) patterns.get(pattern);
		
		Iterator it = candidates.iterator();
		while(it.hasNext())
		{
			String s = (String)it.next();
			Mapping map = new Mapping(word, s);
			
			//System.out.println(s);

			if(isConsistent(currentMap, map))
			{
			//System.out.println(s);							
			//System.out.print("Current Mapping:       ");
			//currentMap.print();
			//System.out.print("New Mapping:           ");
			//map.print();
								
				Mapping mergedMap = mergeKeys(currentMap, map);
				
				//System.out.print("Merged Mapping:        ");
				//map.print();
				
				Mapping newMap = decipher(mergedMap, words);
				
				//System.out.print("Returned Mapping:      ");
				//map.print();
				
				if(validKey(newMap))
				{
					//possibleKeys.add(map);
					currentMap = newMap;
					return currentMap;
				} /*else if(validKey(mergedMap))
				{
					currentMap = mergedMap;
					return currentMap;
				}*/
				else
				{
					possibleKeys.add(currentMap);
					currentMap = preservedMapping;
				}
			}
		}
		if(words.empty() || !words.peek().equals(word))
			words.push(word);
		return currentMap;
	}
	
	/**
	 * validKey()
	 * @param Mapping map
	 * @return bool
	 * @function Validates given key mapping against cipher text and word dictionary
	 */
	public static boolean validKey(Mapping map) {
		String[] cipherWords = globalCipher.split(" ");
		int numCorrect = 0;

		for(int i = 0; i < cipherWords.length; i++)
		{
			String message = decrypt(map, cipherWords[i]);
			if(message.length() == cipherWords[i].length() && dictionary.containsKey(message))
				numCorrect++;
		} 
		
		if(numCorrect == cipherWords.length)
			return true;
		
		return false;
	}
	
	/**
	 * evaluate()
	 * @param Mapping map
	 * @return float score
	 * @function Calculates scoring for a given Mapping to evaluate accuracy
	 */
	public static float evaluate(Mapping map) {
		String[] cipherWords = globalCipher.split(" ");
		float score = 0;

		for(int i = 0; i < cipherWords.length; i++)
		{
			String message = decrypt(map, cipherWords[i]);
			if(message.length() == cipherWords[i].length() && dictionary.containsKey(message))
				score += message.length()*(.35);
		}
		
		// TODO: Add final validation against plain-text sources using recursive substring search/comparisons - if no match is found, key mapping is false
		
		return score;
	}
	
	public static String decrypt(Mapping mapping, String cipher) {
		char[] key = mapping.getKey();
		String plainText = "";
		for(int i = 0; i < cipher.length(); i++)
		{
			if((int)cipher.charAt(i)-(int)'A' < 0)
				plainText += " ";
			else if(key[(int)cipher.charAt(i)-(int)'A'] == (char)0)
				plainText += "*";
			else
				plainText += key[(int)cipher.charAt(i)-(int)'A'];
		}
		
		return plainText;
	}
	
	public static Mapping mergeKeys(Mapping currentMapping, Mapping newMapping) {
		char[] currentKey = currentMapping.getKey();
		char[] newKey = newMapping.getKey();
		
		for(int i = 0; i < 26; i++)
		{
			if(currentKey[i] == (char)0)
				currentKey[i] = newKey[i];
		}
		
		currentMapping.setKey(currentKey);
		
		return currentMapping;
	}
	
	public static Mapping unmerge() {
		return new Mapping();
	}
	
	/**
	 * isConsistent()
	 * @param Mapping currentMapping
	 * @param Mapping newMapping
	 * @return bool
	 * @function Checks Mapping consistency
	 */
	public static boolean isConsistent(Mapping currentMapping, Mapping newMapping) {
		char[] currentKey = currentMapping.getKey();
		char[] newKey = newMapping.getKey();
		
		
		for(int i = 0; i < 26; i++)
		{
			// Check for redundant mappings
			for(int j = 0; j < 26; j++)
			{
				if(newKey[j] == (char)0)
					continue;
				if(newKey[j] == currentKey[i] && i != j)
					return false;
			}
			
			// Check for conflicted mappings
			if(currentKey[i] == (char)0 || newKey[i] == (char)0)
				continue;

			if(currentKey[i] != newKey[i])
				return false;
				

		}

		return true;
	}

	public static class StringLengthComparator implements Comparator<String> {
		public int compare(String o1, String o2) {
			if (o1.length() > o2.length()) {
			  return -1;
			} else if (o1.length() < o2.length()) {
			  return 1;
			} else {
			  return 0;
			}
		}
	}
	
	public static class FrequencyComparator implements Comparator {
		public int compare(Object o1, Object o2) {
			if(dictionary.get(o1) < dictionary.get(o2))
				return -1;
			else if(dictionary.get(o1) > dictionary.get(o2))
				return 1;
			else
				return 0;
		}
	}
	
	/**
	*	parseFile()
	*	@param String fileName
	*	@return void
	*	@function Compiles the given fileName and creates a dictionary of all words 
	*/
	public static void parseFile(String fileName) {
		File srcFile = new File(fileName);

		try {
			Scanner fileScanner = new Scanner(srcFile);
			StringBuilder rawInput = new StringBuilder();
			
			while (fileScanner.hasNextLine()) {
				rawInput.append(fileScanner.nextLine());
				rawInput.append(" ");
			}
			
			String input = rawInput.toString();
		
			input = input.toUpperCase();
			//Replace anything that is not A to Z or space with a space character
			input = input.replaceAll("[^A-Z\\s]", " ");
			//Replace all instances of more than one consecutive whitespace character with a single space character
			input = input.replaceAll("\\s+", " ");
			//Replace last whitespace with nothing
			input = input.replaceAll("$ ", "");

			
			String[] wordList = input.split(" ");
			Arrays.sort(wordList);
			
			for(int i = 0; i < wordList.length; i++)
			{
				//Insert into dictionary
				if(dictionary.containsKey(wordList[i]))
				{
					int count = dictionary.get(wordList[i]);
					dictionary.put(wordList[i], count+1);
				}
				else
				{
					dictionary.put(wordList[i], 1);
				}
			}

			fileScanner.close();
		} catch (IOException e) {
			System.out.println(e);
		}
	}
	
	/**
	 * outputDictionary()
	 * @param
	 * @return void
	 * @function Outputs contents of dictionary file 
	 */
	public static void outputDictionary()
	{
		File outputFile = new File("dictionary.txt");
		
		try{
			BufferedWriter outputWriter = new BufferedWriter(new FileWriter(outputFile));
			
			// Sort dictionary
			TreeMap sortedDictionary = new TreeMap(dictionary);
			Iterator it = sortedDictionary.entrySet().iterator();
			// Output to file
			while (it.hasNext()) {
				Map.Entry entry = (Map.Entry) it.next();
				outputWriter.write(entry.getKey() + " " + entry.getValue());
				outputWriter.newLine();
			}
			
			outputWriter.close();
		} catch(IOException e) {
			System.out.println(e);
		}
	}
	
	/**
	 * pattern()
	 * @param String word
	 * @return StringBuffer patturnBuffer
	 * @function Converts a word into a pattern
	 */
	public static String pattern(String word) {
		StringBuffer patternBuffer = new StringBuffer();
		patternBuffer.setLength(0);
		Mapping patternMapping = new Mapping();
		int count = 0;
		for (int i = 0; i < word.length(); i++) {
			char c = word.charAt(i);
			if(patternMapping.getKey()[(int)c - (int)'A'] == (char)0)
			{
				patternMapping.getKey()[(int)c - (int)'A'] = (char)(count + (int)'A');
				count++;
			}
		}
		for(int i = 0; i < word.length(); i++)
		{
			char c = word.charAt(i);
			char p = patternMapping.getKey()[(int)c - (int)'A'];
			patternBuffer.append(p);
		}
		return patternBuffer.toString();
	}
	
	/**
	 * patternMatches()
	 * @param String word
	 * @return Vector vec
	 * @function Constructs collection of pattern matches for a given word
	 */
	public static Vector patternMatches(String word) {
		String pat = pattern(word);
		Vector vec = (Vector) patterns.get(pat);
		if (vec == null) {
			vec = new Vector();
			vec.add(word);
			patterns.put(pat, vec);
		}
		else if (!vec.contains(word))
		{
			vec.add(word);
			//Collections.sort(vec, new FrequencyComparator());
		}
		return vec;
	}
	
	/**
	 * loadDictionary()
	 * @param String fileName
	 * @return bool
	 * @function Loads existing dictionary file
	 */
	public static boolean loadDictionary(String fileName) {
		File f = new File(fileName);
		try {
			Scanner fileScanner = new Scanner(f);
			String word = "";
			String pattern = "";
			int freq = 0;
			
			while (fileScanner.hasNextLine()) {
				word = fileScanner.nextLine();
				String[] words = word.split(" ");
				word = words[0];
				freq = Integer.parseInt(words[1]);
				wordFreqCount += freq;
				
				if(!dictionary.containsKey(word))
				{
					dictionary.put(word, freq);

					HashMap<String, Integer> matches = (HashMap<String, Integer>)wordsByLength[word.length()];
					if(!matches.containsKey(word))
						((HashMap<String, Integer>)wordsByLength[word.length()]).put(word, freq);
				}

				patternMatches(word);		// Check for pattern matches and add word if necessary
			}
			
			fileScanner.close();
		} catch (IOException e) {
			return false;
		}
		
		return true;
	}

}

File: src/py/Decipher.py
#!/usr/bin/env python
""" generated source for module Decipher1 """

class Decipher(object):
    """ generated source for class Decipher1 """
    dictionary = {}
    patterns = {}
    globalSources = []
    globalCipher = []
    globalCipherWords = []
    globalCipherText = ""
    wordFreqCount = 0
    possibleKeys = []
    wordsByLength = {}

    @classmethod
    def main(cls, args):
        """ generated source for method main """
        #
        # Reading Source Files including your data structures
        #

        # You should do some pre-processing on given text sources and store them in a file so as to not do pre-processing each time.
		# In this part you should read the files and fill the required data structures, like arrays, and assign values to some variables that you will use for finding the plaintext.
        # 		
        cls.dictionary = {}
        cls.patterns = {}
        cls.possibleKeys = []
        cls.wordsByLength = {}
        cls.wordFreqCount = 0

        #
        # 18 is the max word length in sources 1-3
        #
        for i in range(19):
            cls.wordsByLength[str(i)] = {}

        if not cls.loadDictionary("dictionary.txt") or not cls.loadSources(args["sources"]):
            for src in args["sources"]:
                print "Parsing " + src + "..."
                cls.parseDictionary(src)

            cls.outputDictionary()  
            cls.loadDictionary("dictionary.txt")
            cls.loadSources(args["sources"])

        #
        # Reading cipher text from input file
        #
        inputFileName = args["cp"]
        timeLimit = args["tl"]
        
        cls.parseCipherText(inputFileName)

        for cipherText in cls.globalCipher:

            startTime = long(round(time.time()*1000))
            currentTime = long(round(time.time()*1000))

            while currentTime - startTime < timeLimit:
               plainText = cls.beginDecipher(cipherText)
               currentTime = long(round(time.time ()*1000))
               if not plainText == "":
                   break

            endTime = long(round(time.time()*1000))
            totalTime = endTime - startTime

            try:
                outputFile = open("Results.txt", "w+")

                if plainText == "":
                    outPut = ", ".join(["?", str(totalTime)])
                    print(output + "\n")
                    outputFile.write(output)
                else:
                    output = ", ".join([plainText, str(totalTime)])
                    print(output + "\n")
                    outputFile.write(output)

            except IOError as e:
                print e

        print "Deciphering complete..."


    @classmethod
    def loadDictionary(cls, fileName):
        """ loadDictionary(fileName) """
        try:
            f = open(fileName)
            for line in f:
                line = line.split(" ")
                word = line[0]
                freq = int(line[1])
                cls.wordFreqCount += freq
                if not word in cls.dictionary:
                    cls.dictionary[word] = freq
                    matches = {}
                    if not word in matches:
                        cls.wordsByLength[str(len(word))] = (word, freq)
                cls.patternMatches(word)
     
            print fileName, "loaded..." 
        except IOError as e:
            return False
       
        return True


    @classmethod
    def loadSources(cls, fileNames):
        try:
            for f in fileNames:
                with open("parsed-" + f, "r") as srcFile:
                    cls.globalSources.append(srcFile.read())
                    print f, "loaded..."
        except IOError as e:
            return False

        return True


    @classmethod
    def outputDictionary(cls):
        """ outputDictionary() """
        try:
            outputFile = open("dictionary.txt", "w+")
            for key in sorted(cls.dictionary):
                outputFile.write(key + " " + str(cls.dictionary[key]) + "\n")
       
        except IOError as e:
            print e


    @classmethod
    def printDictionary(cls):
        """ printDictionary() """
        for key in sorted(cls.dictionary):
            print key, cls.dictionary[key]


    @classmethod
    def patternMatches(cls, word):
        """ patternMatches(word) """
        pat = cls.pattern(word)
        if not pat in cls.patterns:
            cls.patterns[pat] = []
            cls.patterns[pat].append(word)
        elif not word in cls.patterns[pat]:
            cls.patterns[pat].append(word)


    @classmethod
    def pattern(cls, word):
        """ pattern(word) """
        # Patterns are defined by the word's sequence of unique characters relative to their normed ASCII value
        # Ex 1: HELLO has a pattern mapping of ABCCD
        # Ex 2: ANOMALY has a pattern mapping of ABCDAE
        patternMapping = Mapping()
        count = 0
        
        # Construct new pattern mapping using count to distinguish unique chars
        for ch in word:
            if patternMapping.getKey()[ord(ch) - ord('A')] == "#":
                patternMapping.getKey()[ord(ch) - ord('A')] = chr(count + ord('A'))
                count += 1

        # Construct pattern for word
        pattern = ""
        for ch in word:
            pattern += patternMapping.getKey()[ord(ch) - ord('A')]

        return pattern


    @classmethod
    def parseText(cls, rawInput):
        text = rawInput.upper()
        text = re.sub("\.", "", text)
        text = re.sub("[\W\d_\s]", " ", text)
        text = re.sub("\s+", " ", text)
        text = re.sub(" $", "", text)

        return text


    @classmethod
    def parseDictionary(cls, fileName):
        try:
            with open(fileName, "r") as srcFile:
                rawInput = srcFile.read().replace('\n', '')
            
            fileInput = cls.parseText(rawInput)

            with open("parsed-" + fileName, "w+") as outFile:
                outFile.write(fileInput)

            wordList = fileInput.split(" ")
            for word in wordList:
                if word in cls.dictionary:
                    cls.dictionary[word] += 1
                else:
                    cls.dictionary[word] = 1

        except IOError as e:
            print e


    @classmethod
    def parseCipherText(cls, fileName):
        cipherChunks = []
        try:
            srcFile = open(fileName, "r")
            for line in srcFile:
                rawInput = line.replace('\n', ' ')
                cipher = cls.parseText(rawInput)
                cipherChunks.append(cipher)

            cls.globalCipher = cipherChunks
        
        except IOError as e:
            print e


    @classmethod
    def beginDecipher(cls, cipher):
        plainText = ""
        
        cls.globalCipherText = cipher
        cipherWords = cipher.split(" ")
        cipherWords = sorted(cipherWords, key=len)
        #cipherWords.reverse()

        finalMap = Mapping()
       
        cipherWord = cipherWords.pop()
        pattern = cls.pattern(cipherWord)

        # Find valid mapping
        candidates = cls.patterns[pattern]
        for translation in candidates:
            tempMap = Mapping()
            tempMap.setKey(tempMap.fromTranslation(cipherWord, translation))
            tempMap = cls.decipher(tempMap, cipherWords[:])
            finalMap.setKey(tempMap.getKey()[:])
            if cls.validKey(finalMap):
                break

        # Do validation check on finalMap against other possible mappings that surfaced
        if not cls.validKey(finalMap):
           for keyMapping in cls.possibleKeys:
                possibleMap = Mapping()
                possibleMap.setKey(keyMapping.getKey()[:])
                if cls.evaluate(finalMap) < cls.evaluate(possibleMap):
                    finalMap.setKey(possibleMap.getKey()[:])
        
        plainText = cls.decrypt(finalMap, cipher)
        if not plainText == "":
            print "Final Mapping: ",
            print finalMap.getKey()

        return plainText


    @classmethod
    def decipher(cls, currentMap, cipherWords):
        if not cipherWords:
            return currentMap

        preservedMapping = Mapping()
        preservedMapping.setKey(currentMap.getKey()[:])

        cipherWord = cipherWords.pop()
        pattern = cls.pattern(cipherWord)

        candidates = cls.patterns[pattern]
        for translation in candidates:  
            tempMap = Mapping()
            tempMap.setKey(tempMap.fromTranslation(cipherWord, translation))
            if cls.isConsistent(currentMap, tempMap):
                # Construct consistent merged key mapping
                tempMap = cls.mergeKeys(currentMap, tempMap)
                mergedMap = Mapping()
                mergedMap.setKey(tempMap.getKey()[:])

                # Recursively traverse through remaining cipherWords
                tempMap = cls.decipher(mergedMap, cipherWords[:])
                newMap = Mapping()
                newMap.setKey(tempMap.getKey()[:])
                if cls.validKey(newMap):
                    currentMap.setKey(newMap.getKey()[:])
                    return currentMap
                elif cls.validKey(mergedMap):
                    currentMap.setKey(mergedMap.getKey()[:])
                    return currentMap
                else:
                    # Stores consistent key mappings in case current mapping traversal is incorrect
                    cls.possibleKeys.append(currentMap)
                    currentMap.setKey(preservedMapping.getKey()[:])

        if not cipherWords or not cipherWords[-1] == cipherWord:
            cipherWords.append(cipherWord)
        
        return currentMap

    @classmethod
    def isConsistent(cls, currentMapping, newMapping):
        currentKey = currentMapping.getKey()
        newKey = newMapping.getKey()
        
        for i in range(26):
            if currentKey[i] == "#":
                continue 
            for j in range(26):
                if newKey[j] == "#":
                    continue 
                if newKey[j] == currentKey[i] and i != j:
                    return False
                if not newKey[j] == currentKey[i] and i == j:
                    return False

        return True


    @classmethod
    def mergeKeys(cls, currentMapping, newMapping):
        currentKey = currentMapping.getKey()
        newKey = newMapping.getKey()
 
        for i in range(26):
            if currentKey[i] == "#":
                currentKey[i] = newKey[i]
 
        currentMapping.setKey(currentKey[:])

        return currentMapping


    @classmethod
    def decrypt(cls, mapping, cipher):
        key = mapping.getKey()
        plainText = ""
        
        for ch in cipher:
            if ord(ch) - ord('A') < 0:
                plainText += " "
            elif key[ord(ch) - ord('A')] == "#":
                plainText += "*"
            else:
                plainText += key[ord(ch) - ord('A')]

        return plainText


    @classmethod
    def validKey(cls, mapping):
        numCorrect = 0
        
        cipherWords = cls.globalCipherText.split(" ")
        for cw in cipherWords: 
            message = cls.decrypt(mapping, cw)
            if message in cls.dictionary:
                numCorrect += 1

        if numCorrect == len(cipherWords):
            return True

        return False


    @classmethod
    def evaluate(cls, mapping):
        score = 0

        cipherWords = cls.globalCipherText.split(" ")
        for cw in cipherWords:
            message = cls.decrypt(mapping, cw)
            if len(cw) == len(message) and message in cls.dictionary:
                score += (0.35) * len(message)

        #for cipher in cls.globalCipher:
        #    message = decrypt(mapping, cipher)


        return score


if __name__ == '__main__':
    import sys
    import re
    import argparse
    import time
    from Mapping import *
    
    parser = argparse.ArgumentParser(description="Deciphers a given cipher text with the provided source files.", usage="%(prog)s [options]")
    parser.add_argument("sources", type=str, nargs="+", default="", help="Path to files of original text")
    parser.add_argument("--cp", type=str, nargs="?", default="", help="Path to file containing cipher text", required=True)
    parser.add_argument("--tl", type=long, nargs="?", default=3000L, help="Set time limit for deciphering text in milliseconds")

    args = parser.parse_args()
    args = vars(args)

    # TODO: Add copy constructor to Mapping class

    Decipher.main(args)

